pkg/operators/wasm/wasm.go

```diff
@@ -0,0 +1,235 @@
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/oci"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+)
+
+const (
+	wasmObjectMediaType = "application/vnd.gadget.wasm.program.v1+binary"
+)
+
+type wasmOperator struct{}
+
+func (w *wasmOperator) Name() string {
+	return "wasm"
+}
+
+func (w *wasmOperator) Description() string {
+	return "handles wasm programs"
+}
+
+func (w *wasmOperator) InstantiateImageOperator(
+	gadgetCtx operators.GadgetContext,
+	desc ocispec.Descriptor,
+	paramValues api.ParamValues,
+) (
+	operators.ImageOperatorInstance, error,
+) {
+	return &wasmOperatorInstance{
+		desc:      desc,
+		gadgetCtx: gadgetCtx,
+		handleMap: map[uint32]any{},
+		logger:    gadgetCtx.Logger(),
+	}, nil
+}
+
+type wasmOperatorInstance struct {
+	desc      ocispec.Descriptor
+	rt        wazero.Runtime
+	gadgetCtx operators.GadgetContext
+	mod       wapi.Module
+
+	logger logger.Logger
+
+	// malloc function exported by the guest
+	guestMalloc wapi.Function
+
+	// Golang objects are exposed to the wasm module by using a handleID
+	handleMap  map[uint32]any
+	handleCtr  uint32
+	handleLock sync.RWMutex
+}
+
+func (i *wasmOperatorInstance) Name() string {
+	return "wasm"
+}
+
+func (i *wasmOperatorInstance) Prepare(gadgetCtx operators.GadgetContext) error {
+	err := i.init(gadgetCtx)
+	if err != nil {
+		return fmt.Errorf("initializing wasm: %w", err)
+	}
+	err = i.Init(gadgetCtx)
+	if err != nil {
+		return fmt.Errorf("initializing wasm guest: %w", err)
+	}
+
+	return nil
+}
+
+func (i *wasmOperatorInstance) ExtraParams(gadgetCtx operators.GadgetContext) api.Params {
+	return nil
+}
+
+func (i *wasmOperatorInstance) addHandle(obj any) uint32 {
+	if obj == nil {
+		return 0
+	}
+
+	i.handleLock.Lock()
+	defer i.handleLock.Unlock()
+	i.handleCtr++
+	if i.handleCtr == 0 { // 0 is reserved
+		i.handleCtr++
+	}
+	xctr := 0
+	for {
+		if xctr > 1<<32 {
+			// exhausted; TODO: report somehow
+			return 0
+		}
+		if _, ok := i.handleMap[i.handleCtr]; !ok {
+			// register new entry
+			i.handleMap[i.handleCtr] = obj
+			return i.handleCtr
+		}
+		xctr++
+	}
+}
+
+func (i *wasmOperatorInstance) getHandle(handleID uint32) any {
+	i.handleLock.RLock()
+	defer i.handleLock.RUnlock()
+	return i.handleMap[handleID]
+}
+
+func (i *wasmOperatorInstance) delHandle(handleID uint32) {
+	i.handleLock.Lock()
+	defer i.handleLock.Unlock()
+	delete(i.handleMap, handleID)
+}
+
+func (i *wasmOperatorInstance) init(gadgetCtx operators.GadgetContext) error {
+	ctx := gadgetCtx.Context()
+	rtConfig := wazero.NewRuntimeConfig().
+		WithCloseOnContextDone(true).
+		WithMemoryLimitPages(256) // 16MB (64KB per page)
+	i.rt = wazero.NewRuntimeWithConfig(ctx, rtConfig)
+
+	env := i.rt.NewHostModuleBuilder("env")
+
+	i.addLogFuncs(env)
+	i.addDataSourceFuncs(env)
+	i.addFieldFuncs(env)
+
+	if _, err := env.Instantiate(ctx); err != nil {
+		return fmt.Errorf("instantiating host module: %w", err)
+	}
+
+	wasi_snapshot_preview1.MustInstantiate(ctx, i.rt)
```

Comment:
This needs to be optional for security reasons.

Reply:
I updated it to check the return error instead of panicking. 

Regarding making it optional, it's needed by tinygo (see [this](https://github.com/tetratelabs/wazero/blob/5a8a053bff0ae795b264de9672016745cb842070/examples/basic/add.go#L39-L41). We aren't passing `WithFS` nor other stuff, so the module should be isolated enough. Do you have any specific concerns about the isolation?

Reply:
You're right, I was mixing those up.